Mathlib/MeasureTheory/Function/UnifTight.lean

```diff
@@ -0,0 +1,511 @@
+    ∃ (s : Set α) (_ : MeasurableSet s) (_ : μ s < ∞),
+      snorm (sᶜ.indicator f) p μ ≤ ENNReal.ofReal ε := by
+  -- The proof unwraps `Memℒp f p μ` and applies the analogous result for `lintegral`.
+  -- do some arithmetic that will come in useful
+  have hp_pos := zero_lt_one.trans_le hp_one
+  have hp_nz := hp_pos.ne'
+  have hrp_pos : 0 < p.toReal := ENNReal.toReal_pos hp_nz hp_top
+  have hirp_pos : 0 < 1 / p.toReal := div_pos (by norm_num) hrp_pos
+  have hεp : 0 < ε ^ p.toReal := by simp only [Real.rpow_pos_of_pos, hε]
+  -- decode Memℒp into a.e. strong measurability and finite snorm
+  obtain ⟨haesmf, hsnf⟩ := hf
+  -- transform snorm to lintegral
+  rw [snorm_eq_lintegral_rpow_nnnorm hp_nz hp_top] at hsnf
+  replace hsnf := (rpow_lt_top_iff_of_pos hirp_pos).mp hsnf
+  -- get a.e. measurability for the integrand
+  -- XXX: Why does `AEStronglyMeasurable.ennnorm` only give the weaker AEMeasurable?
+  --      It would make sense to me to use `haesmf.ennnorm.aemeasurable` below.
+  have haemnf := haesmf.ennnorm
+  have haemnpf := haemnf.pow_const p.toReal
+  -- use core result for lintegral (needs only AEMeasurable), the target estimate will be in `hsfε`
+  obtain ⟨s, hms, hμs, hsfε⟩ := lintegral_indicator_compl_le haemnpf hsnf hεp
+  rw [← lintegral_indicator _ hms.compl] at hsfε
+  use s, hms, hμs
+  -- move indicator through function compositions, XXX: is this simp-able?
+  rw [← Function.comp_def (fun x : ℝ≥0∞ => x ^ p.toReal)] at hsfε
+  rw [← Function.comp_def ENNReal.ofNNReal] at hsfε
+  rw [← Function.comp_def nnnorm] at hsfε
+  rw [sᶜ.indicator_comp_of_zero (@ENNReal.zero_rpow_of_pos p.toReal hrp_pos)] at hsfε
+  rw [sᶜ.indicator_comp_of_zero ENNReal.coe_zero] at hsfε
+  rw [sᶜ.indicator_comp_of_zero nnnorm_zero] at hsfε
+  rw [Function.comp_def nnnorm] at hsfε
+  rw [Function.comp_def ENNReal.ofNNReal] at hsfε
+  rw [Function.comp_def (fun x : ℝ≥0∞ => x ^ p.toReal)] at hsfε
+  -- commute ENNReal coersion with rpow, use rpow monotonicity
+  rw [← ofReal_rpow_of_pos (by assumption)] at hsfε
+  rw [← ENNReal.rpow_le_rpow_iff hirp_pos, ENNReal.rpow_inv_rpow_self hrp_pos.ne'] at hsfε
+  -- convert lintegral to snorm
+  rw [← snorm_eq_lintegral_rpow_nnnorm hp_nz hp_top] at hsfε
+  exact hsfε
+
+/-- A constant function is tight. -/
+theorem unifTight_const {g : α → β} (hp : 1 ≤ p) (hp_ne_top : p ≠ ∞) (hg : Memℒp g p μ) :
+    UnifTight (fun _ : ι => g) p μ := by
+  intro ε hε
+  by_cases hε_top : ε = ∞
+  · exact ⟨∅, (by measurability), fun _ => hε_top.symm ▸ le_top⟩
+  have hrε : 0 < ε.toReal := ENNReal.toReal_pos hε.ne' hε_top
+  obtain ⟨s, _, hμs, hgε⟩ := hg.snorm_indicator_compl_le hp hp_ne_top hrε
+  exact ⟨s, ne_of_lt hμs, fun _ => hgε.trans_eq (ENNReal.ofReal_toReal hε_top)⟩
+
+/-- A single function is tight. -/
+theorem unifTight_subsingleton [Subsingleton ι] (hp_one : 1 ≤ p) (hp_top : p ≠ ∞)
+    {f : ι → α → β} (hf : ∀ i, Memℒp (f i) p μ) : UnifTight f p μ := fun ε hε ↦ by
+  by_cases hε_top : ε = ∞
```

Comment:
I see that you consider `ε = ∞` separately here and there. Probably, we should change the definition to `∀ ε : ℝ≥0`, not `∀ ε : ℝ≥0∞`, then add some casts to the first 2 lemmas.

Reply:
I don't see any difference between `ε : ℝ≥0` with `0 < ε` and `ε : ℝ` with `0 < ε` (except that ℝ arithmetic is easier to work with), which is what this PR was using originally. The move to `ε : ℝ≥0∞` did make some intermediate inequalities simpler by getting rid of casts. I think the only reason to go back would be some definite `Mathlib` convention for `ε`, `δ` hypotheses, which so far I haven't noticed.